R/tax_unique.R

```diff
@@ -0,0 +1,279 @@
+#' predefined resolution. In many cases palaeobiologists achieve this by
+#' retaining only unique occurrences identified to the given taxonomic
+#' resolution, however here we also retain occurrences identified to a coarser
+#' resolution which are not already represented within the dataset.
+#' If we take the following set of occurrences:
+#' Albertosaurus sarcophagus
+#' Ankylosaurus sp.
+#' Aves indet.
+#' Ceratopsidae indet.
+#' Hadrosauridae indet.
+#' Ornithominus sp.
+#' Tyrannosaurus rex
+#' then a filter for species-level identifications will reduce the species
+#' diversity to two. However, none of these clades are nested within another, so
+#' each of the indeterminately identified occurrences represents at least one
+#' species not already represented in the dataset. This function is designed to
+#' deal with such taxonomic data, and would retain all seven 'species' in this
+#' dataset.
+#' Data can be supplied in one of two forms, either a dataframe containing
+#' standard occurrence data from the Paleobiology Database, or a series of
+#' vectors which each describe a different taxonomic level for a set of
+#' occurrences (the order of names in the vectors must correspond). At a
+#' minimum, species, genus, and family vectors must be supplied. Occurrence
+#' data can be filtered to retain either unique species or unique genera, and
+#' can be defined as unique either across the whole dataset or within set
+#' categories (e.g. temporal intervals, spatial bins). Missing data should be
+#' indicated with NAs.
+#'
+#' @param paleobioDB \code{dataframe}. A dataframe of taxonomic occurrences
+#' downloaded directly from the Paleobiology Database. The dataframe must
+#' include the following columns: class, order, family, genus, accepted_name.
+#' @param species \code{character}. A vector of species names.
+#' @param genus \code{character}. A vector of genus names.
+#' @param family \code{character}. A vector of family names.
+#' @param order \code{character}. A vector of order names.
+#' @param class \code{character}. A vector of class names.
+#' @param resolution \code{character}. The taxonomic resolution at which to
+#' identify unique occurrences, either species (the default) or genera.
+#' @param by \code{numeric, character}. A category within which to determine
+#' unique taxa. When using \code{paleobioDB}, this should specify the
+#' column within the dataframe to use as the category. When using individual
+#' taxonomic vectors, this should be an additional vector of numbers or
+#' character-based labels designating which category each occurrence belongs to.
+#'
+#' @return A \code{dataframe} of unique taxa, with row numbers corresponding to
+#' the original dataset.
+#'
+#' @section Developer(s):
+#' Bethany Allen
+#' @section Reviewer(s):
+#'
+#' @examples
+#' #Retain unique species
+#' tax_unique(paleobioDB = tetrapods)
+#' tax_unique(species = c("rex", "aegyptiacus", NA),
```

Comment:
There's a few comments on the data input format which I'll try to address collectively here, especially as there seem to be some conflicting ones about adding extra functionality (e.g. here) but also indicating that the function is already too complicated. Essentially I'm unsure as to what non-PBDB dataframes would look like, and we don't have any example data to design around. IMO it's really fundamental that the function can take a PBDB dataframe, because the structure of the accepted_name column is unusual and a real pain to dismantle  - it's the only place to obtain species names from, but the column also includes higher taxonomic names in the case of uncertain identification, so I don't want to place any pre-wrangling burden on users if we can avoid it. For non-PBDB data, I considered entering vectors to be the simplest approach, but I appreciate that it's not in keeping with the rest of the package and requires extra checking for consistent lengths etc. Perhaps the best approach would be to ask users to provide an occdf, then specify whether the dataframe is PBDB, or if not to provide the names of the columns linked to each taxonomic level. I can add the possibility of inputting a binomial column if we are in favour of maximising data flexibility, but this would be at the cost of making the function more complicated.